src/backends/sysfs.rs

```diff
@@ -0,0 +1,1374 @@
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// SPDX-FileCopyrightText: © 2024 Google
+// Ported from libtypec (Rajaram Regupathy <rajaram.regupathy@gmail.com>)
+
+//! The sysfs backend
+
+use mockall_double::double;
+use regex::Regex;
+
+use std::path::Path;
+use std::path::PathBuf;
+
+use crate::pd::PdPdo;
+use crate::ucsi::ConnectorCapabilityOperationMode;
+use crate::ucsi::GetAlternateModesRecipient;
+use crate::ucsi::GetPdoSourceCapabilitiesType;
+use crate::ucsi::GetPdosSrcOrSink;
+use crate::ucsi::PdMessage;
+use crate::ucsi::PdMessageRecipient;
+use crate::ucsi::PdMessageResponseType;
+use crate::ucsi::UcsiAlternateMode;
+use crate::ucsi::UcsiCableProperty;
+use crate::ucsi::UcsiCapability;
+use crate::ucsi::UcsiConnectorCapability;
+use crate::ucsi::UcsiConnectorStatus;
+use crate::BcdWrapper;
+use crate::Error;
+use crate::OsBackend;
+use crate::Result;
+
+#[double]
+use sysfs_reader::SysfsReader;
+#[double]
+use sysfs_walker::SysfsWalker;
+
+const SYSFS_TYPEC_PATH: &str = "/sys/class/typec";
+const SYSFS_PSY_PATH: &str = "/sys/class/power_supply";
+
+/// Creates a `PathBuf` from a string and returns an error if the path does not
+/// exist.
+fn check_path(path: &str) -> Result<PathBuf> {
+    let path = PathBuf::from(path);
+    if !path.exists() {
+        Err(Error::NotSupported {
+            #[cfg(feature = "backtrace")]
+            backtrace: std::backtrace::Backtrace::capture(),
+        })
+    } else {
+        Ok(path)
+    }
+}
+
+/// A module to differentiate `SysfsReader` from `MockSysfsReader`. This is a
+/// limitation of the `mockall` crate.
+pub mod sysfs_reader {
+    #[cfg(test)]
+    use mockall::{automock, predicate::*};
+
+    use std::io;
+    use std::io::Cursor;
+    use std::path::Path;
+    use std::path::PathBuf;
+
+    use crate::pd::Pd3p2BatterySupplyPdo;
+    use crate::pd::Pd3p2DiscoverIdentityResponse;
+    use crate::pd::Pd3p2FastRoleSwap;
+    use crate::pd::Pd3p2FixedSupplyPdo;
+    use crate::pd::Pd3p2SprProgrammableSupplyPdo;
+    use crate::pd::Pd3p2VariableSupplyPdo;
+    use crate::ucsi::CablePropertyPlugEndType;
+    use crate::ucsi::CablePropertyType;
+    use crate::ucsi::ConnectorCapabilityOperationMode;
+    use crate::ucsi::GetPdosSrcOrSink;
+    use crate::ucsi::PdMessageRecipient;
+    use crate::vdo::Pd3p2CertStatVdo;
+    use crate::vdo::Pd3p2IdHeaderVdo;
+    use crate::vdo::Pd3p2ProductTypeVdo;
+    use crate::vdo::Pd3p2ProductVdo;
+    use crate::BcdWrapper;
+    use crate::BitReader;
+    use crate::Error;
+    use crate::FromBytes;
+    use crate::Result;
+
+    use super::SYSFS_TYPEC_PATH;
+
+    /// A mockable sysfs reader.
+    pub struct SysfsReader(Option<PathBuf>);
+
+    #[cfg_attr(test, automock)]
+    impl SysfsReader {
+        pub fn new() -> Result<Self> {
+            Ok(Self(None))
+        }
+
+        pub fn set_path(&mut self, path: &str) -> Result<()> {
```

Comment:
I get that this exists for automock purposes, but it feels *extremely* strange and bug-prone to have mutable calls immediately always followed by a function that consumes the value. IMO this should instead do something like take in a trait responsible for reading files and then have the test code override it to also record all the paths, or mock a `read_file(PathBuf)` instead and just don't check the exact result value. Either way, then you could just thread the path through the actual `read_*` calls.